Revert "Delete lib directory"

DavBot02 opened the pull request:
Reverts DavBot02/World-Clock#1

DavBot02 committed:
Revert "Delete lib directory"

Piinks commented:
cc @goderbauer & @snat-s

goderbauer commented on lib/views/location.dart:
```diff
@@ -0,0 +1,108 @@
+  void initState() {
+    super.initState();
+
+    setUpTime();
```
you also need to implement `didUpdateWidget` to check in there whether `location` has changed. If it has, you will likely have to run setUpTime again...

goderbauer commented on lib/views/location.dart:
```diff
@@ -0,0 +1,108 @@
+
+class _LocationState extends State<Location> {
+  void setUpTime() async {
+    await widget.location.getTime();
```
This `getTime` API is a little strange. I always expect a method names "get" to return something, but this one doesn't. Maybe instead of storing the data internally to the LocationTime return it here as an object that this state can hold on to?

goderbauer commented on lib/views/location.dart:
```diff
@@ -0,0 +1,108 @@
+                      ),
+                      Container(
+                          width: 200,
+                          child: Text(
+                              'Day of the year: ${widget.location.dayOfyear.toString()}')),
```
here an elsewhere: use a trailing commas for some nicer formatting:

```suggestion
                      Container(
                        width: 200,
                        child: Text(
                          'Day of the year: ${widget.location.dayOfyear.toString()}',
                        ),
                      ),
```

goderbauer reviewed (commented):
Congrats on writing your first Flutter app!

Consider adding some widget tests to your app to get a feel for the testing framework. When submitting code to the Flutter framework we always require tests to make sure that a given functionality continues to work and doesn't regress.

Piinks commented on lib/main.dart:
```diff
@@ -0,0 +1,48 @@
+  final Set<LocationTime> favoriteLocations = <LocationTime>{};
+
+  void updateFavoriteLocations(bool isFavorite, LocationTime toFavorite) {
+    setState(() {
+      if (isFavorite) {
+        favoriteLocations.remove(toFavorite);
+      } else {
+        favoriteLocations.add(toFavorite);
+      }
+    });
```
Do you have any hints to better alternatives? I do not know that they are easily discoverable. This is likely a great topic to document for users.

goderbauer replied:
https://docs.flutter.dev/development/data-and-backend/state-mgmt/simple#changenotifier

Piinks commented on lib/routes/home.dart:
```diff
@@ -0,0 +1,42 @@
+                return Location(location: location);
+              }).toList(),
+            )
+          : const Center(child: Text("Add a location!")),
```
Same nit from above here:

```suggestion
          : const Center(child: Text('Add a location!')),
```

Piinks commented on lib/utils/location_time.dart:
```diff
@@ -0,0 +1,41 @@
+
+  Future<void> getTime() async {
+    try {
+      Response response =
+          await get(Uri.parse('http://worldtimeapi.org/api/timezone/$url'));
+      Map data = jsonDecode(response.body);
+
+      String datetime = data['datetime'];
+      String utcOffset = data['utc_offset'].substring(1, 3);
+
+      DateTime now = DateTime.parse(datetime);
```
It looks like all of these could be `final`.

Piinks commented on lib/utils/location_time.dart:
```diff
@@ -0,0 +1,41 @@
+      weekNumber = data['week_number'];
+      time = DateFormat.jm().format(now);
+    } catch (error) {
+      time = "could not get time data";
```
```suggestion
      time = 'could not get time data';
```

Piinks reviewed (commented):
Well done! 🎉 
Do not be alarmed by all of the comments, most are just style nits. Learning the style guide takes time.

Commit:
First version patch

Commit:
First version patch

DavBot02 commented:
Thank you both for your comments, I corrected some style issues and I changed the way the favorite locations are handled -it now uses ChangeNotifier as @goderbauer suggested. I also added a future builder that allows the information to be displayed once it gets fetched. Some functions, variables and files now have a more descriptive name, too.

Let me know if there's anything else that could be worked on

goderbauer commented on analysis_options.yaml:
```diff
@@ -7,7 +7,7 @@
 # The following line activates a set of recommended lints for Flutter apps,
 # packages, and plugins designed to encourage good coding practices.
-include: package:flutter_lints/flutter.yaml
+include: package:lints/recommended.yaml
```
I would recommend to continue using the `flutter_lints/flutter.yaml` here. That file enables some flutter-specific lint rules and includes all the general rules from `lints/recommended.yaml`.

goderbauer replied:
For my own understanding (and to see if we have this documented incorrectly somewhere): What made you change this?

DavBot02 replied:
It was a mistake on my part, for some reason I thought that `flutter_lints/flutter.yaml` was the same as `package:lints/core.yaml` and that `package:lints/recommended.yaml` was a superset.

goderbauer commented on lib/main.dart:
```diff
@@ -0,0 +1,37 @@
+import 'package:flutter/material.dart';
+import 'package:provider/provider.dart';
+import 'routes/home.dart';
```
nit: separate the block of `package:` imports from the block of relative imports by a blank line.

goderbauer commented on lib/main.dart:
```diff
@@ -0,0 +1,37 @@
+            foregroundColor: Colors.black,
+          ),
+        ),
+        initialRoute: '/',
```
replace `/` with the `root` constant?

goderbauer commented on lib/main.dart:
```diff
@@ -0,0 +1,37 @@
+  State<MyApp> createState() => _MyAppState();
+}
+
+class _MyAppState extends State<MyApp> {
```
It looks like this `State` object doesn't manage any state. Could `MyApp` just be a `StatelessWidget`?

goderbauer commented on lib/routes/home.dart:
```diff
@@ -0,0 +1,40 @@
+      body: Consumer<FavoriteLocations>(
+        builder: (context, favoriteLocations, child) {
+          return favoriteLocations.currentLocations.isNotEmpty
+              ? PageView(
```
Consider using the `PageView.builder` constructor here: https://master-api.flutter.dev/flutter/widgets/PageView/PageView.builder.html

The builder constructor lets you instantiate the `LocationCard`s lazily when they are needed instead of instantiating them all upfront even if the user never swipes to them to show them.

goderbauer commented on lib/utils/location_data.dart:
```diff
@@ -0,0 +1,39 @@
+  int? weekNumber;
+  bool? isDayTime;
+
+  Future<bool> fetchData() async {
```
Why is this returning a Future of **bool**?

goderbauer commented on lib/utils/location_data.dart:
```diff
@@ -0,0 +1,39 @@
+
+  String location;
+  String url;
+  String? time;
+  int? dayOfWeek;
+  int? dayOfyear;
+  int? unixTime;
+  int? weekNumber;
+  bool? isDayTime;
```
Anybody, who obtains a `LocationData` object, can currently write to these fields, which is probably not desirable. Can you refactor this, so outsiders can only read these fields and only `fetchData` can modify them?

goderbauer commented on lib/views/location_card.dart:
```diff
@@ -0,0 +1,130 @@
+                                ),
+                              ),
+                            ),
+                            Container(
```
These Containers are all pretty similar. Maybe factor them out into a `DetailsCard` widget, that only takes the parts that a different (e.g. a `title` and a `value`) as parameters?

goderbauer reviewed (commented):
This is looking pretty good! Nice integration of package:provider!
